metadata-io/src/main/java/com/linkedin/metadata/graph/neo4j/Neo4jGraphService.java

```diff
@@ -90,67 +116,177 @@ public void addEdge(@Nonnull final Edge edge) {
+      setupdatedOnTemplate = String.format("r.updatedOn = %s", edge.getUpdatedOn());
+      propertiesTemplateJoiner.add(setupdatedOnTemplate);
+    }
+    if (edge.getUpdatedActor() != null) {
+      setupdatedActorTemplate = String.format("r.updatedActor = '%s'", edge.getUpdatedActor());
+      propertiesTemplateJoiner.add(setupdatedActorTemplate);
+    }
+    if (edge.getProperties() != null) {
+      for (Map.Entry<String, Object> entry : edge.getProperties().entrySet()) {
+        // Make sure extra keys in properties are not preserved
+        final Set<String> preservedKeySet =
+            Set.of("createdOn", "createdActor", "updatedOn", "updatedActor", "startUrn", "endUrn");
+        if (preservedKeySet.contains(entry.getKey())) {
+          throw new UnsupportedOperationException(
+              String.format("Tried setting properties on graph edge but property key is preserved. Key: %s",
+                  entry.getKey()));
+        }
+        if (entry.getValue() instanceof String) {
+          setPropertyTemplate = String.format("r.%s = '%s'", entry.getKey(), entry.getValue());
+          propertiesTemplateJoiner.add(setPropertyTemplate);
+        } else {
+          throw new UnsupportedOperationException(String.format(
+              "Tried setting properties on graph edge but property value type is not supported. Key: %s, Value: %s ",
+              entry.getKey(), entry.getValue()));
+        }
+      }
+    }
+    final String setStartEndUrnTemplate = String.format("r.startUrn = '%s', r.endUrn = '%s'", startUrn, endUrn);
+    propertiesTemplateJoiner.add(setStartEndUrnTemplate);
+    if (!StringUtils.isEmpty(propertiesTemplateJoiner.toString())) {
+      statementR = String.format("%s SET %s", statementR, propertiesTemplateJoiner);
+    }
 
+    statements.add(buildStatement(statement, new HashMap<>()));
+    statements.add(buildStatement(statementR, new HashMap<>()));
     executeStatements(statements);
   }
 
   @Override
   public void upsertEdge(final Edge edge) {
-    throw new UnsupportedOperationException("Upsert edge not supported by Neo4JGraphService at this time.");
+    addEdge(edge);
   }
 
   @Override
   public void removeEdge(final Edge edge) {
-    throw new UnsupportedOperationException("Remove edge not supported by Neo4JGraphService at this time.");
+    log.debug(
+        String.format("Deleting Edge source: %s, destination: %s, type: %s", edge.getSource(), edge.getDestination(),
+            edge.getRelationshipType()));
+
+    final String sourceType = edge.getSource().getEntityType();
+    final String destinationType = edge.getDestination().getEntityType();
+    final String sourceUrn = edge.getSource().toString();
+    final String destinationUrn = edge.getDestination().toString();
+
+    String endUrn = destinationUrn;
+    String startUrn = sourceUrn;
+    String endType = destinationType;
+    String startType = sourceType;
+    String realType = "r_" + edge.getRelationshipType();
```

Comment:
What exactly is `realType`?

Reply:
change realType to reverseRelationshipType
Reverse incoming/outgoing direction check by compare sourceType and relationshipType to LineageSpec.
for example:
sourceType: dataset, relationshipType: downstreamOf.
downstreamOf relationship type and outgoing relationship direction for dataset from LineageSpec, is inside upstreamEdges.
source(dataset) -[downstreamOf]-> dest means upstreamEdge for source(dataset)
dest -[r_downstreamOf]-> source(dataset), need reverse source and dest

sourceType: datajob, relationshipType: produces.
produces relationship type and outgoing relationship direction for datajob from LineageSpec, is inside downstreamEdges.
source(datajob) -[produces]-> dest means downstreamEdge for source(datajob)
source(dataset) -[r_produces]-> dest, do not need to reverse source and dest

reverseRelationshipType means reverse/real relationship between source node and dest node.